deeppavlov/core/data/dataset.py

```diff
@@ -34,24 +51,35 @@ def __init__(self, data: Dict[str, List[Tuple[Any, Any]]], seed: int = None, *ar
             'all': self.train + self.test + self.valid
         }
 
-    def batch_generator(self, batch_size: int, data_type: str = 'train') -> Generator:
-        r"""This function returns a generator, which serves for generation of raw (no preprocessing such as tokenization)
+    def batch_generator(self, batch_size: int, data_type: str = 'train',
+                        shuffle: bool = None, seed: int = None) -> Generator:
+        r"""This function returns a generator, which serves for generation of raw
+        (no preprocessing such as tokenization)
          batches
         Args:
             batch_size (int): number of samples in batch
             data_type (str): can be either 'train', 'test', or 'valid'
+            shuffle (bool): whether to shuffle dataset before batching
+            seed (int): random seed for batching
         Returns:
             batch_gen (Generator): a generator, that iterates through the part (defined by data_type) of the dataset
         """
+        if shuffle is not None:
+            self.shuffle = shuffle
+        if seed is not None:
+            self.seed = seed
+
         data = self.data[data_type]
         data_len = len(data)
         order = list(range(data_len))
-
-        rs = random.getstate()
-        random.setstate(self.random_state)
-        random.shuffle(order)
-        self.random_state = random.getstate()
-        random.setstate(rs)
+        if self.shuffle:
+            rs = random.getstate()
+            random.seed(self.seed)
```

Comment:
Every epoch will most likely have the same seed, so the shuffling is always the same so there is basically no shuffling